packages/headless/src/features/commerce/common/actions.ts

```diff
@@ -34,10 +34,11 @@ export interface QueryCommerceAPIThunkReturn {
 }
 
 export const buildCommerceAPIRequest = async (
-  state: ListingAndSearchStateNeededByQueryCommerceAPI
+  state: ListingAndSearchStateNeededByQueryCommerceAPI,
+  relay: Relay
```

Comment:
Why do we add `relay` as a function parameter (and thus, impact all upstream code calling the function, as we can see in this PR) instead of just `createRelay(....extractStuffFromTheState)` directly in the body of this function ?

Reply:
Or, another alternative, add a new function in the relay package that does only the clientId extraction, without creating a full client.

Reply:
And then, more generally, should this be replicated to **all** API request builder function across the project, when analytics mode is set to `next`. ie: making sure the fix is correctly applied to all use cases, not just commerce.

Reply:
> Or, another alternative, add a new function in the relay package that does only the clientId extraction, without creating a full client.

I kind of like this idea.

On the other hand, I don't think there's anything wrong with what we're doing right now: the Relay client is initialized once in the core engine, and then passed down as needed.

Reply:
> On the other hand, I don't think there's anything wrong with what we're doing right now:

What I don't like is the "shotgun effect" it creates in the codebase (one change === multiple file changes).

"Passing it down as needed" is complexity to figure out where something actually comes from/is created (multiple level upstream), vs just a local variable/function (easier to reason about).

If would be one thing if the builder function was really lacking data to get what it needs from relay, but from my understanding it's already all available, so it's a bit silly IMO to do it like this PR propose.

Reply:
Not opposed to it, but I'd like to see a design from the perspective of an implementer first and then we can decide on the next steps. There's a similar discussion happening [here](https://github.com/coveo/ui-kit/pull/4113#discussion_r1653008709),